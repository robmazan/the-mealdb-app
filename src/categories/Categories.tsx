import React, { useEffect, useState } from "react";
import { Category, fetchCategories } from "../api/theMealDb";
import CategoryThumbnail from "./CategoryThumbnail";
import styles from "./Categories.module.scss";

enum LoadingState {
  PENDING,
  ERROR,
  DONE
}

const Categories: React.FC = () => {
  const [categories, setCategories] = useState<Category[]>([]);
  const [loadingState, setLoadingState] = useState<LoadingState>(
    LoadingState.PENDING
  );

  useEffect(() => {
    const fetchIfNeeded = async () => {
      let storedCategories = localStorage.getItem("storedCategories");
      let categories: Category[] = [];

      if (storedCategories) {
        categories = JSON.parse(storedCategories);
        setLoadingState(LoadingState.DONE);
      } else {
        try {
          categories = await fetchCategories();
          localStorage.setItem("storedCategories", JSON.stringify(categories));
          setLoadingState(LoadingState.DONE);
        } catch {
          setLoadingState(LoadingState.ERROR);
        }
      }
      setCategories(categories);
    };
    fetchIfNeeded();
  }, []);

  switch (loadingState) {
    case LoadingState.DONE:
      return (
        <ul className={styles.list}>
          {categories.map(category => (
            <li key={category.idCategory} className={styles.listItem}>
              <CategoryThumbnail category={category} />
            </li>
          ))}
        </ul>
      );

    case LoadingState.ERROR:
      return <div>ERROR: cannot load categories!</div>;

    case LoadingState.PENDING:
      return <div>Loading categories...</div>;
  }
};

export default Categories;
