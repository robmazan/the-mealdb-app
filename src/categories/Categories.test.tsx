import { GlobalWithFetchMock, MockResponseInit } from "jest-fetch-mock";
import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import Categories from "./Categories";

const customGlobal: GlobalWithFetchMock = global as GlobalWithFetchMock;
customGlobal.fetch = require("jest-fetch-mock");
const fetchMock = customGlobal.fetch;

jest.mock("./CategoryThumbnail");

let container: HTMLElement | null = null;
beforeEach(() => {
  fetchMock.resetMocks();
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  if (container) {
    unmountComponentAtNode(container);
  }
  container!.remove();
  container = null;
  localStorage.clear();
});

it("shows loading message", async () => {
  const mockPromise = new Promise<MockResponseInit>(resolve => {});

  fetchMock.mockResponseOnce(() => mockPromise);

  await act(async () => {
    render(<Categories />, container);
  });

  expect(container!.innerHTML).toContain("Loading");
});

it("renders category", async () => {
  const fakeCategory = require("./__mocks__/fakeCategory.json");
  const fakeResponse = {
    categories: [fakeCategory]
  };
  fetchMock.mockResponseOnce(JSON.stringify(fakeResponse));

  await act(async () => {
    render(<Categories />, container);
  });

  const mockHTMLContent = container!.querySelector("code");

  if (mockHTMLContent) {
    const renderedCategory = JSON.parse(mockHTMLContent.innerHTML);
    expect(renderedCategory).toMatchObject(fakeCategory);
  } else {
    fail(
      `The mocked <CategoryThumbnail> has no <code> element that should contain the serialized JSON for the category! Container: ${
        container!.innerHTML
      }`
    );
  }
});

it("shows error message on fetch error", async () => {
  fetchMock.mockRejectOnce(new Error("Fetch error"));

  // Use the asynchronous version of act to apply resolved promises
  await act(async () => {
    render(<Categories />, container);
  });

  expect(container!.innerHTML).toContain("ERROR");
});

it("caches categories", async () => {
  const fakeCategory = require("./__mocks__/fakeCategory.json");
  const fakeResponse = {
    categories: [fakeCategory]
  };
  fetchMock.mockResponseOnce(JSON.stringify(fakeResponse));

  await act(async () => {
    render(<Categories />, container);
  });

  // For testing re-fetch we need a new container, otherwise
  // ReactDOM wouldn't re-render as the component output is the same
  const otherContainer = document.createElement("div");
  document.body.appendChild(otherContainer);

  // shouldn't re-fetch on second render but use the cached value
  await act(async () => {
    render(<Categories />, otherContainer);
  });

  expect(fetchMock.mock.calls.length).toEqual(1);

  unmountComponentAtNode(otherContainer);
  otherContainer.remove();
});
