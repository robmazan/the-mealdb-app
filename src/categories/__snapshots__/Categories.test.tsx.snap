// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders category 1`] = `
<Fragment>
  <h2
    className="pageTitle"
  >
    Categories
  </h2>
  <ul
    className="list"
  >
    <li
      className="listItem"
      key="1"
    >
      <Thumbnail
        imgAlt="Mock category"
        imgHeight={125}
        imgSrc="https://example.com/thumb.jpg"
        imgWidth={200}
        label="Mock category"
        to="/category/Mock category"
        tooltip="Mock description"
      />
    </li>
  </ul>
</Fragment>
`;
